packages/core-data/src/reducer.js

```diff
@@ -214,6 +289,66 @@ export const entities = ( state = {}, action ) => {
 	};
 };
 
+/**
+ * Reducer keeping track of entity edit undo history.
+ *
+ * @param {Object} state  Current state.
+ * @param {Object} action Dispatched action.
+ *
+ * @return {Object} Updated state.
+ */
+const UNDO_INITIAL_STATE = [];
+UNDO_INITIAL_STATE.offset = 0;
+export function undo( state = UNDO_INITIAL_STATE, action ) {
+	switch ( action.type ) {
+		case 'EDIT_ENTITY_RECORD':
+			if ( action.meta.isUndo || action.meta.isRedo ) {
+				const nextState = [ ...state ];
+				nextState.offset = state.offset + ( action.meta.isUndo ? -1 : 1 );
+				return nextState;
+			}
+
+			// Transient edits don't create an undo level, but are
+			// reachable in the next meaningful edit to which they
+			// are merged. They are defined in the entity's config.
+			if ( ! Object.keys( action.edits ).some( ( key ) => ! action.transientEdits[ key ] ) ) {
```

Comment:
Kind of related to the previews comment, but we could decide to only set the meta undo key if there are non-transient edits.

Reply:
But we still need transient edits to always go through this so that they are applied/unapplied.